Broadcast message functionality

kchadha opened the pull request:
### Resolves

Resolves parts of #749 ("Broadcast Menu new broadcast doesn't do anything" and "Broadcast menu doesn't show broadcasts except the preset ones")

### Proposed Changes

This pull request adds the following pieces of functionality:
- All of the broadcast message related blocks are now populated with two options, "message1" which is a preset message the user can broadcast, and "New message..." which, like "Rename Variable"/"Delete Variable" for lists and variables will bring up a modal prompting the user to create a new message.
- Messages are treated as typed variables with the string "broadcast_msg" differentiating this type of variable from lists and scalar variables
- The default option of "message1" gets added to the variable model with the correct type (specified above) when the block gets added to the workspace.
- When the new message option is selected and the user is able to successfully create a new message, that message gets added to the variable model and the field on the block that was used to create the new message gets updated with the new message name.
- When a new message is created, all broadcast message drop-downs get updated with the new message name as a menu option.

### Reason for Changes

Broadcast messages were previously non-functional.

### Test Coverage

The changes have been tested with the existing unit tests.
@rachel-fenichel If you think there are any specific tests that need to be changed or added, please let me know.

### Additional Comments

There are a few outstanding tasks/issues remaining. These are outlined below:

- Modify the messages in the modals that pop up when creating or renaming lists and broadcast messages (issue #1244)
- Allow messages, lists, and variables to share common names (issue #1245). Resolving this issue might also require changing how variables are looked up, because I believe there are certain places where variables are looked up by name (type information would not get properly checked here).
- A next step (not covered in this PR) is to allow the drop-down field for the 'broadcast' and 'broadcast and wait' blocks to be input fields (allowing dropping other blocks into that slot).
- A question that came up is: we are currently are specifying broadcast messages as variables by specifying that their field dropdown is specifically a FieldVariable. We also specify that the drop downs should only contain variables of the types specified in the variableTypes list. E.g. broadcast message block drop downs should only contain names of broadcast messages (and not scalar variables or lists). Often, we are using this variableTypes list to discern what type the actual variable we're trying to create (by means of the 'New message...' option in the dropdown) is. See proposed changes to FieldVariable.initModel and FieldVariable.dropdownCreate, for examples. What should happen if variableTypes has more than one type in it? Perhaps we need a different way of specifying in the json representation of the block what the type of the variable is (similar to how we define a default message name for the broadcast message dropdowns, via 'variable: "message1"'

kchadha committed:
Adding ability to create  a new broadcast message using the dropdown option. Broadcast messages are now typed variables with the variable type as 'broadcast_msg'. Needed to do some extra work to translate variables from blocks to XML correctly because the default broadcast messages were getting added as new variables with the incorrect type upon trying to move any of the broadcast message related blocks from the toolbar into the workspace because it was trying to create a new variable with a different type than already existed. Currently, broadcast message names conflict with variable names, and I need to make it so that isn't so. Additionally, need to update the source block's message name after creation of a new message via the dropdown.

kchadha committed:
If a field dropdown is used to create a new message, the field then gets updated with the name of the newly created message. There are also some todo comments here for future work that has been documented in two github issues (#1244 and #1245)
.

kchadha committed:
Make one of the comments clearer.

kchadha committed:
Another comment clarification.

rachel-fenichel commented:
https://i.imgur.com/jUTHMLn.gif

I'm excited to look through this.

For variable dropdowns with multiple types, a single "new variable" option would be insufficient.  In that case I would argue for having an explicit "Create variable" button in the flyout per type of variable, or a modal that lets you pick the type.  This is one of the very underspeced parts of typed variables.

rachel-fenichel commented on core/field_variable.js:
```diff
@@ -81,7 +81,18 @@ Blockly.FieldVariable.prototype.initModel = function() {
+    // a scalar variable
+    if (this.getVariableTypes_().length === 1) {
+      this.sourceBlock_.workspace.createVariable(this.getValue(),
+        this.getVariableTypes_()[0]);
```
I expected that you would add an explicit type to the scalar variables ('scalar', maybe).  What's the advantage of leaving it as the empty string?

For context, here's `workspace.createVariable`: 
```
/**
 * Create a variable with a given name, optional type, and optional id.
 * @param {!string} name The name of the variable. This must be unique across
 *     variables and procedures.
 * @param {?string} opt_type The type of the variable like 'int' or 'string'.
 *     Does not need to be unique. Field_variable can filter variables based on
 *     their type. This will default to '' which is a specific type.
 * @param {?string} opt_id The unique id of the variable. This will default to
 *     a UUID.
 * @return {?Blockly.VariableModel} The newly created variable.
 */
Blockly.Workspace.prototype.createVariable = function(name, opt_type, opt_id) {
  if (name.toLowerCase() == Blockly.Variables.noVariableText()) {
    return;
  }
  return this.variableMap_.createVariable(name, opt_type, opt_id);
};
```

kchadha replied:
I agree, we should call out scalar variables specifically instead of leaving the original empty string to specify that type. The only reason that the change hasn't been made in this PR is that it seemed like it would be a very pervasive change and out of the scope of this PR, which is specifically for adding functionality behind the dropdown options for the broadcast messages. I will file an issue for this change.

kchadha replied:
This is now referenced in issue #1251

rachel-fenichel commented on msg/messages.js:
```diff
@@ -148,6 +148,10 @@ Blockly.Msg.CANNOT_DELETE_VARIABLE_PROCEDURE = 'Can\'t delete the variable "%1"
 
+// Broadcast Message creation
+/// dropdown choice - Create a new message.
+Blockly.Msg.NEW_MESSAGE = 'New message...';
```
Let's call this "NEW_BROADCAST_MESSAGE".  Msg.NEW_MESSAGE will be confusing down the line.

rachel-fenichel commented on blocks_vertical/event.js:
```diff
@@ -77,13 +77,10 @@ Blockly.Blocks['event_whenbroadcastreceived'] = {
-            ['new message', 'new message']
-          ]
+          "variableTypes": ["broadcast_msg"],
+          "variable": "message1"
```
Is "message1" going to be the same for all languages?  Blockly usually uses messages here so that they can be translated.

For instance, the default name [here](https://developers.google.com/blockly/guides/create-custom-blocks/variables#untyped_variable_blocks) is given by "BKY_VARIABLES_DEFAULT_NAME"

kchadha committed:
Changing new message constant name to be more useful because Blockly.Msg.NEW_MESSAGE will probably get confusing.

kchadha committed:
Making default name for broadcast messages amenable to translation.

kchadha committed:
Merge branch 'develop' into broadcast-message-functionality

kchadha committed:
Merge branch 'develop' into broadcast-message-functionality

kchadha commented:
@rachel-fenichel Thank you for looking over this PR and for your comments! I just pushed some changes addressing your comments about message names for translation. I also filed a new issue to address your comment regarding calling out the type for scalar variables more specifically (rather than keeping it as an empty string). Please let me know if there are any other changes to be made. Thanks again!

rachel-fenichel commented on core/field_variable.js:
```diff
@@ -171,13 +182,17 @@ Blockly.FieldVariable.dropdownCreate = function() {
     // doesn't modify the workspace's list.
     for (var i = 0; i < variableTypes.length; i++) {
       var variableType = variableTypes[i];
+      if (variableType === 'broadcast_msg'){
```
Make 'broadcast_msg' a constant, probably in constants.js.

kchadha replied:
Should I also use == here instead of ===?

rachel-fenichel replied:
Yes.

rachel-fenichel commented on core/xml.js:
```diff
@@ -116,6 +116,21 @@ Blockly.Xml.blockToDom = function(block, opt_noId) {
+          // FieldVariableGetter doesn't have getVariableTypes_ function
+          if (field instanceof Blockly.FieldVariable) {
+            var variableTypes = field.getVariableTypes_();
+            if (variableTypes.length === 1) {
```
use == instead of === here and above.

(We've had the long discussion about whether == or === is more correct and decided that both can lead to bugs, but we've settled on == unless === is specifically called for.)

rachel-fenichel reviewed (approved):
Looks good overall.  Two style changes, and then feel free to merge.

For looking up variables by id instead of name: yup, that's a real problem, and it's upstream: https://github.com/google/blockly/issues/1199.  It's also the cause of bad change events for variables.  I'm looking into it for now.  

You don't need to try to fix it, but leave a note anywhere that it's problematic so that I can go through and do a cleanup once I get it right in Blockly.  You can refer to https://github.com/LLK/scratch-blocks/issues/1253

The last attempt to fix it failed and had to be rolled back: https://github.com/google/blockly/pull/1254

kchadha committed:
Refactoring 'broadcast_msg' to be referenced via a constant in core/constants instead of being hard-coded everywhere.

kchadha committed:
Changing === to ==

kchadha committed:
Adding some TODO comments referencing existing issues also discussed in this PR.

kchadha commented:
@rachel-fenichel just letting you know that I added some TODO comments referencing #1245 and #1253 as part of this PR in my most recent commit.
